sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapp/AADOAuth2UserService.java

```diff
@@ -115,20 +118,39 @@ public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2Authenticatio
         if (!properties.allowedGroupIdsConfigured() && !properties.allowedGroupNamesConfigured()) {
             return Collections.emptySet();
         }
-        Set<String> groups = Optional.of(accessToken)
-                                     .map(AbstractOAuth2Token::getTokenValue)
-                                     .map(graphClient::getGroupsFromGraph)
-                                     .orElseGet(Collections::emptySet);
-        if (properties.getUserGroup().getEnableFullList()) {
-            return groups;
+        Map<String, Set<String>> groups = getGroups(accessToken);
+
+        if (properties.getUserGroup().getEnableFullList() || isAllGroupId()) {
+            Set<String> allowedGroupsName = Arrays
+                .asList(properties.getUserGroup().getAllowedGroupNames())
+                .stream()
+                .flatMap(Collection::stream)
+                .filter(groups.get(GROUP_NAME)::contains)
+                .collect(Collectors.toSet());
+            return Arrays.asList(groups.get(GROUP_ID), allowedGroupsName)
+                         .stream()
+                         .flatMap(Collection::stream)
+                         .map(group -> ROLE_PREFIX + group)
+                         .collect(Collectors.toSet());
         }
-        Set<String> roles = Arrays
+        return Arrays
             .asList(properties.getUserGroup().getAllowedGroupIds(), properties.getUserGroup().getAllowedGroupNames())
             .stream()
             .flatMap(Collection::stream)
-            .filter(groups::contains)
+            .filter(group -> groups.get(GROUP_ID).contains(group) || groups.get(GROUP_NAME).contains(group))
             .map(group -> ROLE_PREFIX + group)
             .collect(Collectors.toSet());
-        return roles;
+    }
+
+    private boolean isAllGroupId() {
```

Comment:
1. `isAllGroupId` -> `allGroupIdShouldMapToRole`
2. Move `properties.getUserGroup().getEnableFullList()` into this method.